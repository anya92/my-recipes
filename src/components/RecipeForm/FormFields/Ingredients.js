import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import Label from '../styled/Label';
import Input from '../styled/Input';

const Ingredient = styled.div`
  display: flex;
  margin: 4px 0;
  height: 40px;
  ${Input} {
    width: calc(100% / 2 - 20px);
    flex: 1 1;
    &:first-child {
      ${props => (props.i === 0 && props.error) && 'border: 2px solid crimson;'};
    }
    &:nth-child(2) {
      margin-left: 4px;
    }
    &::-webkit-input-placeholder {
      color: #C4C4C4;
      font-size: 1.1rem;
    }
  }
`;

const Delete = styled.span`
  background-color: #F2F2F2;
  font-size: 1.6rem;
  flex-basis: 40px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  padding: 10px;
  margin-left: 4px;
  cursor: pointer;
`;

const Add = styled.div`
  width: 100%;
  height: 40px;
  margin-top: 5px;
  background-color: #F2F2F2;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
`;

const Ingredients = ({
  ingredients, handleIngredientChange, addIngredientInput, removeIngredientInput, error,
}) => (
  <div>
    <Label>Ingredients *</Label>
    {
      ingredients.map((ing, i) => (
        <Ingredient key={i} error={error} i={i}>
          <Input
            placeholder="name"
            value={ingredients[i].name}
            onChange={e => handleIngredientChange(e, 'name', i)}
          />
          <Input
            placeholder="quantity"
            value={ingredients[i].quantity}
            onChange={e => handleIngredientChange(e, 'quantity', i)}
          />
          <Delete onClick={() => removeIngredientInput(i)}>&times;</Delete>
        </Ingredient>
      ))
    }
    <Add onClick={addIngredientInput}>Add new ingredient</Add>
  </div>
);

Ingredients.propTypes = {
  ingredients: PropTypes.arrayOf(PropTypes.object).isRequired,
  handleIngredientChange: PropTypes.func.isRequired,
  addIngredientInput: PropTypes.func.isRequired,
  removeIngredientInput: PropTypes.func.isRequired,
  error: PropTypes.bool.isRequired,
};

export default Ingredients;
