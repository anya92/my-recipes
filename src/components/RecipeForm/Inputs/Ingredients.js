import React from 'react';
import styled from 'styled-components';

const StyledIngredients = styled.div`
  font-size: 1.2rem;
  label {
    display: block;
    text-transform: capitalize;
    color: #4F4F4F;
    font-weight: 700;
    margin: 5px 0;
  }
`;

const Ingredient = styled.div`
  display: flex;
  margin: 4px 0;
  height: 40px;
  input {
    border: none;
    background-color: #F2F2F2;
    padding: 8px 2px;
    width: calc(100% / 2 - 20px);
    flex: 1 1;
    border-bottom: 2px solid #F2F2F2;
    &:focus {
      border-bottom-color: dodgerblue;
      outline: none;
    }
    &:first-child {
      ${props => (props.i === 0 && props.error) && 'border: 2px solid crimson;'};
    }
    &:nth-child(2) {
      margin-left: 4px;
    }
    &::-webkit-input-placeholder {
      color: #C4C4C4;
      font-size: 1.1rem;
    }
  }
`;

const Delete = styled.span`
  background-color: #F2F2F2;
  font-size: 1.6rem;
  flex-basis: 40px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  padding: 10px;
  margin-left: 4px;
  cursor: pointer;
`;

const Add = styled.div`
  width: 100%;
  height: 40px;
  margin-top: 5px;
  background-color: #F2F2F2;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
`;

const Ingredients = ({
  ingredients, handleIngredientChange, addIngredientInput, removeIngredientInput, error,
}) => (
  <StyledIngredients>
    <label htmlFor="ingredients">Ingredients</label>
    {
      ingredients.map((ing, i) => (
        <Ingredient key={i} error={error} i={i}>
          <input
            placeholder="name"
            value={ingredients[i].name}
            onChange={e => handleIngredientChange(e, 'name', i)}
          />
          <input
            placeholder="quantity"
            value={ingredients[i].quantity}
            onChange={e => handleIngredientChange(e, 'quantity', i)}
          />
          <Delete onClick={() => removeIngredientInput(i)}>&times;</Delete>
        </Ingredient>
      ))
    }
    <Add onClick={addIngredientInput}>Add new ingredient</Add>
  </StyledIngredients>
);

export default Ingredients;
