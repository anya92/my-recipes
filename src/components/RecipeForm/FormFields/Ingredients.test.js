import React from 'react';
import { shallow } from 'enzyme';
import recipes from '../../../recipes';

import Ingredients from './Ingredients';

let ingredients;
let handleIngredientChange;
let addIngredientInput;
let removeIngredientInput;
let error;
let wrapper;

beforeEach(() => {
  ingredients = [{ name: '', quantity: '' }, { name: '', quantity: '' }, { name: '', quantity: '' }];
  handleIngredientChange = jest.fn();
  addIngredientInput = jest.fn();
  removeIngredientInput = jest.fn();
  error = false;
  wrapper = shallow(<Ingredients
    ingredients={ingredients}
    handleIngredientChange={handleIngredientChange}
    addIngredientInput={addIngredientInput}
    removeIngredientInput={removeIngredientInput}
    error={error}
  />);
});

test('should render Ingredients with no ingredients correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render Ingredients with ingredients correctly', () => {
  wrapper.setProps({ ingredients: recipes[0].ingredients });
  expect(wrapper).toMatchSnapshot();
});

test('should call addIngredientInput on Add click', () => {
  wrapper.find('AddIngredient').simulate('click');
  expect(addIngredientInput).toHaveBeenCalled();
});

test('should call removeIngredientInput on Delete click', () => {
  wrapper.find('DeleteIngredient').at(0).simulate('click');
  expect(removeIngredientInput).toHaveBeenLastCalledWith(0);
});

test('should call handleIngredienChange on input change', () => {
  const name = 'honey';
  const quantity = '1 tablespoon';
  const nameInput = wrapper.find('Ingredient').at(0).find('Input').at(0);
  const quantityInput = wrapper.find('Ingredient').at(0).find('Input').at(1);
  nameInput.simulate('change', { target: { value: name } });
  expect(handleIngredientChange).toHaveBeenLastCalledWith({ target: { value: name } }, 'name', 0);
  quantityInput.simulate('change', { target: { value: quantity } });
  expect(handleIngredientChange).toHaveBeenLastCalledWith({ target: { value: quantity } }, 'quantity', 0);
});
