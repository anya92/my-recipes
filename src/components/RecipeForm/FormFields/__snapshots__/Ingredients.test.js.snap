// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Ingredients with ingredients correctly 1`] = `
<div>
  <styled.label>
    Ingredients *
  </styled.label>
  <Ingredient
    error={false}
    i={0}
    key="0"
  >
    <Input
      onChange={[Function]}
      placeholder="name"
      value="eggs"
    />
    <Input
      onChange={[Function]}
      placeholder="quantity"
      value="2-4"
    />
    <DeleteIngredient
      onClick={[Function]}
    >
      ×
    </DeleteIngredient>
  </Ingredient>
  <Ingredient
    error={false}
    i={1}
    key="1"
  >
    <Input
      onChange={[Function]}
      placeholder="name"
      value="butter"
    />
    <Input
      onChange={[Function]}
      placeholder="quantity"
    />
    <DeleteIngredient
      onClick={[Function]}
    >
      ×
    </DeleteIngredient>
  </Ingredient>
  <Ingredient
    error={false}
    i={2}
    key="2"
  >
    <Input
      onChange={[Function]}
      placeholder="name"
      value="salt"
    />
    <Input
      onChange={[Function]}
      placeholder="quantity"
    />
    <DeleteIngredient
      onClick={[Function]}
    >
      ×
    </DeleteIngredient>
  </Ingredient>
  <Ingredient
    error={false}
    i={3}
    key="3"
  >
    <Input
      onChange={[Function]}
      placeholder="name"
      value="pepper"
    />
    <Input
      onChange={[Function]}
      placeholder="quantity"
    />
    <DeleteIngredient
      onClick={[Function]}
    >
      ×
    </DeleteIngredient>
  </Ingredient>
  <AddIngredient
    onClick={[Function]}
  >
    Add new ingredient
  </AddIngredient>
</div>
`;

exports[`should render Ingredients with no ingredients correctly 1`] = `
<div>
  <styled.label>
    Ingredients *
  </styled.label>
  <Ingredient
    error={false}
    i={0}
    key="0"
  >
    <Input
      onChange={[Function]}
      placeholder="name"
      value=""
    />
    <Input
      onChange={[Function]}
      placeholder="quantity"
      value=""
    />
    <DeleteIngredient
      onClick={[Function]}
    >
      ×
    </DeleteIngredient>
  </Ingredient>
  <Ingredient
    error={false}
    i={1}
    key="1"
  >
    <Input
      onChange={[Function]}
      placeholder="name"
      value=""
    />
    <Input
      onChange={[Function]}
      placeholder="quantity"
      value=""
    />
    <DeleteIngredient
      onClick={[Function]}
    >
      ×
    </DeleteIngredient>
  </Ingredient>
  <Ingredient
    error={false}
    i={2}
    key="2"
  >
    <Input
      onChange={[Function]}
      placeholder="name"
      value=""
    />
    <Input
      onChange={[Function]}
      placeholder="quantity"
      value=""
    />
    <DeleteIngredient
      onClick={[Function]}
    >
      ×
    </DeleteIngredient>
  </Ingredient>
  <AddIngredient
    onClick={[Function]}
  >
    Add new ingredient
  </AddIngredient>
</div>
`;
